Chapter 5 restructure (#136)
chhoumann/p7: report/sections/Architecture/architecture.tex

This is done by executing tests. The database component is used to persist account information, test runs, and syllabus data.
In the following sections we elaborate on the choices made for each component, in relation to this structure.
The overall architecture for the system can be seen in figure \ref{fig:Architecture}.

> The database and test runner runs as processes contained in different docker containers. The frontend and backend components run in the same container.

Reviewer: ```suggestion
The database and test runner run as processes contained in different Docker containers. The frontend and backend components run in the same container.
```

The remaining entities \textit{Account}, \textit{Session} and \textit{VerificationToken} are entities which are required by authentication services and are not developed by us.

\section{Backend System}
The backend component has several responsibilities including routing, page rendering, authentication, role based data access and modification, as well as communication between other components. 
The component enables the creation of accounts for the system by using an access provider.

> For this project Google is used as this provider. 

Reviewer: ```suggestion
For this project, Google is used as this provider. 
```

> Once a user account is created they are assigned the role either of lecturer or student.  

Reviewer: ```suggestion
Once a user account is created, they are assigned the role either of lecturer or student.  
```

Depending on the account role, the user has access to different functionality.

> A lecturer is able to create and modify syllabi, problemsets, submissions, problems and tests. 

Reviewer: ```suggestion
A lecturer is able to create and modify syllabi, problem sets, submissions, problems, and tests. 
```

> A student is able to fetch syllabi, problemsets, problems and tests, but is not able to create or modify them. 

Reviewer: ```suggestion
A student is able to fetch syllabi, problem sets, problems, and tests, but is not able to create or modify them. 
```

> A student is able to create submissions.

Reviewer: ```suggestion
A student is able to create submissions through tRPC.
```

> This is done using tRPC. 

Reviewer: ```suggestion
```

> Once the component receives a query or mutation the input is validated to ensure that it is of correct structure.

Reviewer: ```suggestion
Once the component receives a query or mutation, the input is validated to ensure that it is of correct structure.
```

If the query or mutation is formatted correctly, and the user has the correct authorization, the Prisma ORM transforms it to database queries, and maps the result to \javascript{} objects.
The backend then renders a response page or object and sends it to the user.
The component is implemented in TypeScript and Next.js.

\section{Frontend} \label{sec:architecture-frontend}

> The frontend component is the user facing component of our platform.  It contains the UI for the platform and is the main interface for the system.

Reviewer: ```suggestion
The frontend component is the user facing component of our platform.  It contains the user interface for the platform and is the main interface of the system.
```

The component is designed such that its user interface and functionality is dependent on the role of the user. 

> A user's role is stored in the form of an encrypted cookie which sent to the backend whenever mutations or queries are made. 

Reviewer: ```suggestion
A user's role is stored in the form of an encrypted cookie, which sent to the backend whenever mutations or queries are made. 
```

> The component is also responsible for providing users only with UI elements they are authorized to view. 

Reviewer: ```suggestion
The component is also responsible for only providing users with user interface elements they are authorized to view. 
```

For instance, some edit and delete buttons are available only to lecturers.

> The component contains functionality providing users with syntax highlight using the CodeMirror library. 

Reviewer: ```suggestion
The component contains functionality providing users with syntax highlighting using the CodeMirror library. 
```

It is implemented using TypeScript and React. 

\section{Test Runner}

> The test runner component is responsible for determining whether a student's submission solves a problem. 

Reviewer: ```suggestion
The test runner component is responsible for determining whether a student exercise submissions solves a problem.
```

> It receives HTTP requests containing Haskell code and an Hspec test, and use this to stage and perform test execution. 

Reviewer: ```suggestion
It receives HTTP requests containing Haskell code and an Hspec test, and uses this to stage and perform test execution. 
```

This is done using the GHC interpreter.

> The result of the test is then serialized to JSON format and sent to the backend component.

Reviewer: ```suggestion
The result of processing the test is then serialized to JSON and sent to the backend component.
```

> The test runner component is implemented as an HTTP server using the Axum crate in the Rust programming language.

Reviewer: ```suggestion
The test runner component is implemented as an HTTP server using the axum crate in the Rust programming language.
```
